Integrate http tunnel

MakMuftic opened the pull request:
<!--
Thank you for your pull request. Please provide all required information
-->

**Short description of work done**
<!-- e.g. Refactored auth logic as part of the transition to OAuth -->
Integrate and customize  [go-http-tunel](https://github.com/mmatczuk/go-http-tunnel) inside repo. Used [fork](https://github.com/datadope-io/go-http-tunnel/tree/added_port_auto_assignation) with autoassign port on server side.

### PR Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [X] I have run linter localy
- [X] I have run unit and integration tests locally
- [X] Rebased to master branch / merged master

### Changes
<!-- Please describe all changes made to codebase. -->
- Refactored auth to use token inside `X-Auth-Header`
- Restructured code and expose new (simpler) functions for creating `Client` and `Server`
- Refactor logging to be compatible with the rest of the repo
- Remove unnecessary code such as cmd commands from the original repo
- Refactor tests
- Tidy `go.mod` file

### ~~Example~~
<!-- You can add screenshots or videos to show changed behaviour -->
#### ~~Creating server~~
```go
s, err := server.NewServer(&server.ServerConfig{
	TlsCrtFilePath: "./pkg/http-tunel/testdata/selfsigned.crt",
	TlsKeyFilePath: "./pkg/http-tunel/testdata/selfsigned.key",
	Address:        ":5223",
	PortRange:      "10000:50000",
	AuthHandler: func(s string) bool {
		return s == "test-token"
	},
	Logger: log.NewEntry(l),
})

if err != nil {
	l.Fatal("failed to create server: %s", err)
}

s.Start()
```
#### ~~Creating client~~
```go
c, err := client.NewClient(&client.ClientConfig{
	ServerAddress:  "127.0.0.1:5223",
	TlsCrtFilePath: "./pkg/http-tunel/testdata/selfsigned.crt",
	TlsKeyFilePath: "./pkg/http-tunel/testdata/selfsigned.key",
	Tunnels: map[string]*client.Tunnel{
		"": {
			Protocol:   "tcp",
			Addr:       "localhost:3000",
			Auth:       "",
			Host:       "",
			RemoteAddr: "0.0.0.0:AUTO",
		},
	},
	Logger:    log.NewEntry(l),
	AuthToken: "test-token",
})

if err != nil {
	l.Fatal("failed to create client: %s", err)
}

if err := c.Start(); err != nil {
	l.Fatal("failed to start tunnels: %s", err)
}
```

### Issues
<!-- Use github keyword to close issues that are related to this PR -->

<!-- [REQUIRED] -->
Closes #51 - issue
Closes #36 - original discussion

Commit:
Extracted http-tunel code and refactored logging

Commit:
Add util functions for running server/client http-tunel

Commit:
Restructure code to server and client packages

Commit:
Refactor integration test

Commit:
Fix integration test

Commit:
Remove doc file

Commit:
Fix control message

Commit:
Fix fetching local address

Commit:
Refactor server creation function for http tunel

Commit:
Code cleanup

Commit:
Remove test cert and clean up go.mod file

Commit:
Tidy mod file

Commit:
Remove tunnel initialization from load balancer

Commit:
Code cleanup on server

Commit:
Fix linter errors

Commit:
Rename dir name

Commit:
Fix linter errors

Commit:
Bump go version

Commit:
Add windows build directive

mpetrun5 commented on pkg/http-tunnel/client/client.go:
```diff
@@ -0,0 +1,377 @@
+		MaxTime:     DefaultBackoffMaxTime,
+	})
+
+	clientData.idName = "" // TODO
```
?

mpetrun5 commented on pkg/http-tunnel/client/client.go:
```diff
@@ -0,0 +1,377 @@
+}
+
+func newClient(config *clientData) (*Client, error) {
+	if config.serverAddr == "" {
```
Does it make sense to check all of these fields below again when we checked them or assigned them in NewClient?

mpetrun5 commented on pkg/http-tunnel/client/client.go:
```diff
@@ -0,0 +1,377 @@
+	w.Header().Add("X-Auth-Header", c.config.authToken)
+	w.WriteHeader(http.StatusOK)
+
+	// te := TunnelExt{
```
?

mpetrun5 commented on pkg/http-tunnel/proto/controlmsg.go:
```diff
@@ -0,0 +1,88 @@
+	if msg.ForwardedProto == "" {
+		missing = append(missing, HeaderForwardedProto)
+	}
+	//if msg.ForwardedId == "" {
```
Same as above.
If we don't need this logic might as well erase it.

mpetrun5 commented on pkg/http-tunnel/server/addrpool.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+type RemoteID struct {
+	//ConnID   id.ID
```
?

mpetrun5 commented on pkg/http-tunnel/server/pool.go:
```diff
@@ -0,0 +1,139 @@
+	}
+}
+
+/*func (p *connPool) URL(identifier id.ID) string {
```
?

mpetrun5 commented on pkg/http-tunnel/server/pool.go:
```diff
@@ -0,0 +1,139 @@
+func (p *connPool) GetClientConn(req *http.Request, addr string) (*http2.ClientConn, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	//fmt.Printf("GET CLIENNNNNNNNT CONNNNNNNN---------%s : ", addr)
```
Client goes BRRRRRRRRRRRRRRRRRR :joy:

mpetrun5 commented on pkg/http-tunnel/server/pool.go:
```diff
@@ -0,0 +1,139 @@
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	//addr := p.addr(identifier)
```
?

mpetrun5 commented on pkg/http-tunnel/server/pool.go:
```diff
@@ -0,0 +1,139 @@
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	//addr := p.addr(identifier)
```
?

mpetrun5 commented on pkg/http-tunnel/server/registry.go:
```diff
@@ -0,0 +1,245 @@
+	}
+	reg := r.source[origaddr]
+	reg.ClientName = cname
+	//fmt.Printf("SUBSCRIBE REGISTRY Client Name: [%s] Client ID: [%s] value : %+v \n", cname, origaddr, reg)
```
Remove

mpetrun5 commented on pkg/http-tunnel/server/registry.go:
```diff
@@ -0,0 +1,245 @@
+	if !ok {
+		return "", nil, false
+	}
+	fmt.Printf("SUBSCRIBER REGISTRY [%s] value : %+v \n", hostPort, h)
```
All of these commented and uncommented prints in this file should be logs or removed

mpetrun5 commented on pkg/http-tunnel/server/registry.go:
```diff
@@ -0,0 +1,245 @@
+		"identifier": identifier,
+	})
+
+	//i, ok := r.items[identifier]
```
There is a lot of unnecessary comented out code that we should look into and erase

mpetrun5 commented on pkg/http-tunnel/server/server.go:
```diff
@@ -0,0 +1,762 @@
+		return nil, fmt.Errorf("listener failed: %s", err)
+	}
+
+	logger := serverData.logger
```
Same as client don't think we need theses as we did the same thing in NewClient
IMO in one place this logic is extra

mpetrun5 commented on pkg/http-tunnel/server/tcpproxy.go:
```diff
@@ -0,0 +1,146 @@
+
+func (p *TCPProxy) localAddrFor(hostPort string) string {
+
+	//	fmt.Printf("TCPPROXY localAddrFor GET FROM %s: %#+v\n ", hostPort, p.localAddrMap)
```
Someone really liked commented out prints

Commit:
Refactor build tags

Commit:
Fix function name

Commit:
Revert build tags

mpetrunic reviewed (commented):
- What's up with tls configs, can we remove those?
- Can you move tunnel logs from info to debug?

Commit:
Remove unnecessary params check from client and code cleanup

Commit:
Remove all comented code

Commit:
Code cleanup server

Commit:
Remove comented code block

Commit:
Fix failing test

Commit:
WIP remove tls connection

Commit:
WIP remove tls props from config structs

mpetrunic commented on pkg/http-tunnel/client/client.go:
```diff
@@ -0,0 +1,334 @@
+package client
+
+import (
+	"crypto/tls"
```
khm?

Commit:
WIP add server initialization to loadbalancer

Commit:
Remove tls connection from http-tunnel

Commit:
Add integration test

Commit:
Replace all info logging with debug inside http-tunnel

Commit:
Code cleanup

MakMuftic commented:
@mpetrun5 @mpetrunic 
- removed tls connection
- moved all tunnel logs to debug
- added integration test (starts `client` and `server` and checks if the tunnel opened successfully)

### [NEW] Example
#### Creating server
```go 
s, err := server.NewServer(&server.ServerConfig{
	Address:        ":5223",
	PortRange:      "10000:50000",
	AuthHandler: func(s string) bool {
		return s == "test-token"
	},
	Logger: log.NewEntry(l),
})
if err != nil {
	l.Fatalf("failed to create server: %s", err)
}
s.Start()
```
#### Creating client
```go 
c, err := client.NewClient(&client.ClientConfig{
	ServerAddress:  "127.0.0.1:5223",
	Tunnels: map[string]*client.Tunnel{
		"": {
			Protocol:   "tcp",
			Addr:       "localhost:3000",
			Auth:       "",
			Host:       "",
			RemoteAddr: "0.0.0.0:AUTO",
		},
	},
	Logger:    log.NewEntry(l),
	AuthToken: "test-token",
})

if err != nil {
	l.Fatalf("failed to create client: %s", err)
}

if err := c.Start(); err != nil {
	l.Fatalf("failed to start tunnels: %s", err)
}
```

Commit:
Tidy go mod file

mpetrunic commented on pkg/http-tunnel/client/utils.go:
```diff
@@ -0,0 +1,79 @@
+	tcpAddr := make(map[string]string)
+
+	for v, t := range m {
+		fmt.Printf("Creating proxy for %#+v/%#+v\n", v, t)
```
use logger?

mpetrunic commented on pkg/http-tunnel/client/utils.go:
```diff
@@ -0,0 +1,79 @@
+			}
+			httpURL[t.Host] = u
+		case proto.TCP, proto.TCP4, proto.TCP6:
+			fmt.Printf("Setting config for %s | REMOTE: %s | LOCAL: %s\n", v, t.RemoteAddr, t.Addr)
```
ditto

Commit:
Code cleanup
